### PR TITLE
Ocr 114

### PR DESCRIPTION
**Pull request by BenMacnaughton**
Please ensure you've followed the checklist and provide all the required information *before* requesting a review.

## Required Information

- [x] Reference the issue this is meant to fix or describe it if none exists.
- [x] Full description of changes made.

## Pull Request Checklist

- [x] Coding standards are followed (e.g., [PEP8](https://pep8.org/))
- [x] An entry exists for the container in `docker-compose.yml` or `build.yml`
- [x] The Docker image builds
- [x] The container runs correctly when sent inputs with the changes
- [x] No models or other large files are part of these commits
- [x] A description of the tests already run as part of this PR is present

### CODE REVIEW

File: preprocessors/ocr/ocr.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+# and our Additional Terms along with this program.
+# If not, see
+# <https://github.com/Shared-Reality-Lab/IMAGE-server/LICENSE>.
+
+
+import json
+import time
+import logging
+import jsonschema
+import requests
```
Comment (aidanwilliams09) [inline preprocessors/ocr/ocr.py:22]: ```requests``` module not used

File: preprocessors/ocr/ocr.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+import os
+import io
+import base64
+from flask import Flask, request, jsonify
+from azure.cognitiveservices.vision.computervision import ComputerVisionClient
+from azure.cognitiveservices.vision.computervision.models import OperationStatusCodes
+from azure.cognitiveservices.vision.computervision.models import VisualFeatureTypes
```
Comment (aidanwilliams09) [inline preprocessors/ocr/ocr.py:29]: ```VisualFeatureTypes``` module not used

File: preprocessors/ocr/ocr.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+from msrest.authentication import CognitiveServicesCredentials
+
+app = Flask(__name__)
+
+
+@app.route('/preprocessor', methods=['POST', 'GET'])
+def get_ocr_text():
```
Comment (aidanwilliams09) [inline preprocessors/ocr/ocr.py:36]: Helper method and api method are called the same thing. Consider disambiguating them

File: preprocessors/ocr/ocr.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    # Grab the ID from the URL
+    operation_id = read_operation_location.split("/")[-1]
+
+    # Call the "GET" API and wait for it to retrieve the results 
+    # Might get stuck here - need a break condition of some sort
+    while True:
+        read_result = computervision_client.get_read_result(operation_id)
+        if read_result.status not in ['notStarted', 'running']:
+            break
+        time.sleep(1)
```
Comment (aidanwilliams09) [inline preprocessors/ocr/ocr.py:132]: Is there a specific reason we need to sleep after calling the api? Also do we know for a fact that every single piece of data will return some status other than these two at some point?

File: preprocessors/ocr/ocr.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    subscription_key = os.environ["AZURE_API_KEY"]
+    endpoint = "https://image-cv.cognitiveservices.azure.com/"
+    
+    computervision_client = ComputerVisionClient(endpoint, CognitiveServicesCredentials(subscription_key))
+    
+    read_response = computervision_client.read_in_stream(stream,  raw=True)
+    
+    read_operation_location = read_response.headers["Operation-Location"]
+    # Grab the ID from the URL
+    operation_id = read_operation_location.split("/")[-1]
```
Comment (aidanwilliams09) [inline preprocessors/ocr/ocr.py:124]: Is it possible to do a regex here? I don't know whether there's a real benefit to it but splitting and getting a specific index is not very error proof.

File: preprocessors/ocr/ocr.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    return response
+
+def get_ocr_text(source):
+    """
+    Gets OCR text data from Azure API
+    """
+
+    # Convert URI to binary stream
+
+    image_b64 = source.split(",")[1]
```
Comment (aidanwilliams09) [inline preprocessors/ocr/ocr.py:111]: See comment on line 124.

File: preprocessors/ocr/ocr.py (hunk 5 above)
Comment (BenMacnaughton) [reply preprocessors/ocr/ocr.py:124]: This was the recommendation from the Azure docs. I am not attached to it if you feel like changes are needed

File: preprocessors/ocr/ocr.py (hunk 4 above)
Comment (BenMacnaughton) [reply preprocessors/ocr/ocr.py:132]: This was recommended by Azure docs. We do not so I think some sort of time threshold break condition would be suitable

File: preprocessors/ocr/ocr.py (hunk 6 above)
Comment (BenMacnaughton) [reply preprocessors/ocr/ocr.py:111]: This is fine. Request schema ensures that this will work

**Comment by BenMacnaughton** [comment]
Tested on several images via curl. Data returns as expected. No end-to-end testing with browser has been done.

**Comment by BenMacnaughton** [comment]
The corresponding schemas PR for the OCR has already been merged

**Review by jaydeepsingh25 — APPROVED** [review]

**Pull request merged**